### PR TITLE
oop-object_interactions-medium-q4

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/oop/object_interactions/medium/q4/QuestionText.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
+# oop-object_interactions-medium-q4
+
+- Java question
+
+## Question Text
+
+You are working as a vet assistant in an animal hospital, and sometimes you encounter an animal that requires a blood
+transfusion in order to save its life. Suppose you were given the following test class. Make a class called **Snail**
```
Comment (SamamthaH) [inline src/oop/object_interactions/medium/q4/QuestionText.md:8]:
```suggestion
transfusion in order to save its life. Make a class called **Snail**
```

File: src/oop/object_interactions/medium/q4/QuestionText.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    - A `char` called **bloodType**
+    - A `double` called **volume**
+    - None of these attributes should be accessible from other classes.
+
+- Constructors:
+    - A constructor that take a `char` and a `double` to initialize the **bloodType** and **volume** attributes.
+    - It is accessible from other classes.
+
+- Methods:
+    - Getter ans Setter methods for **bloodType** and **volume**, and  **toString** method that prints the object
```
Comment (SamamthaH) [inline src/oop/object_interactions/medium/q4/QuestionText.md:28]:
```suggestion
    - Getter and Setter methods for **bloodType**, **volume**, and  **toString** method that prints the object
```

File: src/oop/object_interactions/medium/q4/QuestionText.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      information.
+    - These methods are accessible from other classes.
+    -
```
Comment (SamamthaH) [inline src/oop/object_interactions/medium/q4/QuestionText.md:31]:
```suggestion
```

File: src/oop/object_interactions/medium/q4/QuestionText.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+You must add the following items to the class:
+
+- Methods:
+    - A `transfusionFrom` method that take another **Snail** object to do a blood tansfer.
+    - It should check if the blood types are a match and if not print `Blood type incompatible`.
```
Comment (SamamthaH) [inline src/oop/object_interactions/medium/q4/QuestionText.md:37]: should probably make a note here that snail blood types are not like humans, and to check if they match exactly or something, since human blood types are a bit more complicated for who can transfer blood to another person and we want to clear up any confusion

File: src/oop/object_interactions/medium/q4/QuestionText.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    - If the types are compatible it should say `Blood transfusion starting...`
```
Comment (SamamthaH) [inline src/oop/object_interactions/medium/q4/QuestionText.md:38]: it feels a bit odd to have this message and then a message saying the transfusion cannot happen, maybe switch around when these statements get printed

File: src/oop/object_interactions/medium/q4/QuestionText.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    - Then it should make sure that the donor has more than 5 unit of blood if not
+      it should print `Donor does not have enough blood for a transfusion`.
+    - If donor has enough take 1 unit of blood from donor
+      and give it to the snail, and then print `Blood transfusion completed`.
+
+## TestTransfusion Class
+
+Your `TestNewspaper` class must have the following items:
```
Comment (SamamthaH) [inline src/oop/object_interactions/medium/q4/QuestionText.md:46]:
```suggestion
Your `TestTransfusion` class must have the following items:
```

File: src/oop/object_interactions/medium/q4/QuestionText.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+- Main Method:
+    - A `main` method which initializes three **Snail** objects called Sarah, Jones, and Cindy who respectively have
+      blood type A, B, and B, and blood volume of 10.5, 8, and 0.5.
+    - Cindy should try to get blood from first Sarah and then Jones.
+    - Then it should print the information of Sarah, Jones, and Cindy in that order using **toString()** method.
+    - An example of the output would be `"Sarah: " + sarah.toString()`.
```
Comment (SamamthaH) [inline src/oop/object_interactions/medium/q4/QuestionText.md:53]: this should also mention if each output is on a separate line

File: src/oop/object_interactions/medium/q4/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+            _assertEquals(volume + 1, testClass.getFieldValue(classInstance, volumeAttributeName),
+                    "Your " + transfusionFromMethodName + " method does not correctly add blood to the recipient.");
+            _assertEquals(donVolume - 1, donor.getVolume(),
+                    "Your " + transfusionFromMethodName + " method does not correctly take blood from the donor.");
+
+        }
+        assertNull(output, String.join(" ", "Your", transfusionFromMethodName, "should not return any output"));
+    }
+
+    @ParameterizedTest
```
Comment (SamamthaH) [inline src/oop/object_interactions/medium/q4/MainTest.java:150]: doesn't the question say that the toString is given? I mean, it makes sense that it wouldn't be, but either keep this test and change the question text or get rid of this test and keep the question text unchanged

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/object_interactions/medium/q4/QuestionText.md (hunk 5 above)
Comment (KaiKhademi) [reply src/oop/object_interactions/medium/q4/QuestionText.md:37]: I think the purpose here was to make students write nested if else statements which it lose that if that changes.

Comment (SamamthaH) [reply src/oop/object_interactions/medium/q4/QuestionText.md:39]: If we just change the contents of the message then that element is still retained. Change it to something like "Checking if donor has enough blood for a transfusion."

File: src/oop/object_interactions/medium/q4/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+                String.format("You have modified the provided %s attribute. Please revert it to the original state.", bloodTypeAttributeName));
+        assertTrue(testClass.hasField(volumeAttributeName, double.class, new String[]{"private"}),
+                String.format("You have modified the provided %s attribute. Please revert it to the original state.", volumeAttributeName));
+        Class<?>[] classArguments = {char.class, double.class};
+        assertTrue(testClass.hasConstructor(classArguments),
+                "You have modified the provided constructor on the " + className + " class. Please revert it to the original state.");
+        assertTrue(testClass.hasModifier(classArguments, "public"),
+                "You have modified the provided constructor on the " + className + " class. Please revert it to the original state.");
+        assertTrue(testClass.hasMethod(getBloodTypeMethodName, null, char.class, new String[]{"public"}),
+                String.format("You have modified the provided %s method method. Please revert it to the original state.", getBloodTypeMethodName));
```
Comment (SamamthaH) [inline src/oop/object_interactions/medium/q4/MainTest.java:53]: get rid of the duplicate "method" in this and the other error messages

File: src/oop/object_interactions/medium/q4/MainTest.java [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+            Object classInstance = testClass.createInstance(arguments);
+            _assertEquals(type, testClass.getFieldValue(classInstance, bloodTypeAttributeName),
+                    "You have modified the provided constructor on the " + className + " class. Please revert it to the original state.");
+            _assertEquals(volume, testClass.getFieldValue(classInstance, volumeAttributeName),
+                    "You have modified the provided constructor on the " + className + " class. Please revert it to the original state.");
+            char updateType = (char) test[num][2];
+            double updateVolume = (double) test[num][3];
+            Object output = testClass.callMethod(getBloodTypeMethodName, classInstance);
+            _assertEquals(type, output,
+                    "You have modified the provided " + getBloodTypeMethodName + "method  on the " + className + " class. Please revert it to the original state.");
```
Comment (SamamthaH) [inline src/oop/object_interactions/medium/q4/MainTest.java:81]:
```suggestion
                    "You have modified the provided " + getBloodTypeMethodName + "method on the " + className + " class. Please revert it to the original state.");
```

Comment (SamamthaH) [reply src/oop/object_interactions/medium/q4/MainTest.java:81]: get rid of this additional space in the other messages as well

File: src/oop/object_interactions/medium/q4/MainTest.java [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+            output = testClass.callMethod(getVolumeMethodName, classInstance);
+            _assertEquals(volume, output,
+                    "You have modified the provided " + getVolumeMethodName + "method  on the " + className + " class. Please revert it to the original state.");
+            output = testClass.callMethod(setBloodTypeMethodName, new Object[][]{{updateType, char.class}}, classInstance);
+            _assertEquals(updateType, testClass.getFieldValue(classInstance, bloodTypeAttributeName),
+                    "You have modified the provided " + setBloodTypeMethodName + "method  on the " + className + " class. Please revert it to the original state.");
+            assertNull(output, String.join(" ", "Your", setBloodTypeMethodName, "should not return any output"));
```
Comment (SamamthaH) [inline src/oop/object_interactions/medium/q4/MainTest.java:88]: this is already checked in the .hasMethod above

File: src/oop/object_interactions/medium/q4/MainTest.java [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+            output = testClass.callMethod(setVolumeMethodName, new Object[][]{{updateVolume, double.class}}, classInstance);
+            _assertEquals(updateVolume, testClass.getFieldValue(classInstance, volumeAttributeName),
+                    "You have modified the provided " + setVolumeMethodName + "method  on the " + className + " class. Please revert it to the original state.");
+            assertNull(output, String.join(" ", "Your", setVolumeMethodName, "should not return any output"));
```
Comment (SamamthaH) [inline src/oop/object_interactions/medium/q4/MainTest.java:92]: this is already checked in the .hasMethod above

File: src/oop/object_interactions/medium/q4/MainTest.java [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+        Object[][] arguments = {
+                {type, char.class},
+                {volume, double.class}
+        };
+        Snail donor = new Snail(donType, donVolume);
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] methodArguments = {
+                {donor, Snail.class}
+        };
+        Object output;
```
Comment (SamamthaH) [inline src/oop/object_interactions/medium/q4/MainTest.java:121]: use a .hasMethod before the if statements with the .callMethod. This will also get rid of the need for an output variable as assertNull will not need to be checked

File: src/oop/object_interactions/medium/q4/MainTest.java [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+        String[] methodModifiers = {"public"};
+        String expected = "My blood type is " + type + " and I have " + volume + "L of blood left";
+        String incorrectToStringMessage = String.join(" ",
+                "Your", className, " toString method does not return the correct String.");
+        Object output = testClass.callMethod("toString", methodModifiers, classInstance);
+        _assertEquals(expected, output, incorrectToStringMessage);
+    }
+
+    @Test
+    @Tag("dependent1")
```
Comment (SamamthaH) [inline src/oop/object_interactions/medium/q4/MainTest.java:170]: This should be tagged "dependency1" according to Seth's example

File: src/oop/object_interactions/medium/q4/MainTest.java [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+                Arguments.of('A', 10.3, 'A', minBlood - 1),
+                Arguments.of('A', 10.3, 'A', minBlood + 1),
+                Arguments.of('B', 4.2, 'A', 2.5),
+                Arguments.of('O', 3.4, 'B', 5.7)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    @Tag("dependent2")
```
Comment (SamamthaH) [inline src/oop/object_interactions/medium/q4/MainTest.java:110]: This should be tagged "dependent1" according to Seth's example

File: src/oop/object_interactions/medium/q4/QuestionText.md [hunk 16]
```diff
@@ -0,0 +1,55 @@
+# oop-object_interactions-medium-q4
+
+- Java question
+
+## Question Text
+
+You are working as a vet assistant in an animal hospital, and sometimes you encounter an animal that requires a blood
+transfusion in order to save its life. Make a class called **Snail**
+with attributes **bloodType** and **volume** that are initialized by the constructor. This class has a method called
```
Comment (SamamthaH) [inline src/oop/object_interactions/medium/q4/QuestionText.md:9]: fix formatting here and get rid of duplicate class information

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/object_interactions/medium/q4/QuestionText.md [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+    - A `toString` method that prints the object information in the format
+      `My blood type is x and I have yL of blood left` where x is the **bloodType** value and y is the **volume** value.
+    - A `transfusionFrom` method that take another **Snail** object to do a blood tansfer.
+    - It should check if the blood types are a match and if not print `Blood type incompatible`. Note that in this scenario
+  **Snail** blood types are not like humans, so you should check that the type match exactly.
+    - If the types are compatible it should say `Checking if donor has enough blood for a transfusion...`
+    - Then it should make sure that the donor has more than 5 unit of blood if not
+      it should print `Donor does not have enough blood for a transfusion`.
+    - If donor has enough take 1 unit of blood from donor
+      and give it to the snail, and then print `Blood transfusion completed`.
```
Comment (Hedgemon4) [inline src/oop/object_interactions/medium/q4/QuestionText.md:34]: These need to be tabbed over, they are not currently sub-bullets under the `transfusion` method.

File: src/oop/object_interactions/medium/q4/QuestionText.md [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+
+You must add the following items to the class:
+
+- Methods:
+    - A `toString` method that prints the object information in the format
+      `My blood type is x and I have yL of blood left` where x is the **bloodType** value and y is the **volume** value.
+    - A `transfusionFrom` method that take another **Snail** object to do a blood tansfer.
+    - It should check if the blood types are a match and if not print `Blood type incompatible`. Note that in this scenario
+  **Snail** blood types are not like humans, so you should check that the type match exactly.
+    - If the types are compatible it should say `Checking if donor has enough blood for a transfusion...`
```
Comment (Hedgemon4) [inline src/oop/object_interactions/medium/q4/QuestionText.md:30]:
```suggestion
    - If the types are compatible it should print `Checking if donor has enough blood for a transfusion...`
```

File: src/oop/object_interactions/medium/q4/QuestionText.md (hunk 17 above)
Comment (Hedgemon4) [reply src/oop/object_interactions/medium/q4/QuestionText.md:34]: Also, specify if all the print statements should be on different lines.

File: src/oop/object_interactions/medium/q4/MainTest.java [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+            output = testClass.callMethod(setVolumeMethodName, new Object[][]{{updateVolume, double.class}}, classInstance);
+            _assertEquals(updateVolume, testClass.getFieldValue(classInstance, volumeAttributeName),
+                    "You have modified the provided " + setVolumeMethodName + " method on the " + className + " class. Please revert it to the original state.");
+            assertNull(output, String.join(" ", "Your", setVolumeMethodName, "should not return any output"));
+        }
+    }
+
+    private static Stream<Arguments> inputProvider() {
+        return Stream.of(
+                Arguments.of('A', 10.3, 'B', 4.9),
```
Comment (Hedgemon4) [inline src/oop/object_interactions/medium/q4/MainTest.java:98]: You should not always use 10.3 and A here, the test cases should be more varied.

Comment (Hedgemon4) [reply src/oop/object_interactions/medium/q4/MainTest.java:98]: Also, please add another test case where the recipient has less blood than the donor but the transfusion occurs.

File: src/oop/object_interactions/medium/q4/MainTest.java [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+        };
+        Object output;
+        if (type != donType) {
+            output = testClass.callMethod(transfusionFromMethodName, methodArguments, new String[]{"public"}, classInstance, new Clause[]{
+                    new StringLiteral("Blood type incompatible"),
+                    new Optional(new StringLiteral(" "))
+            });
+        } else if (donVolume <= minBlood) {
+            output = testClass.callMethod(transfusionFromMethodName, methodArguments, new String[]{"public"}, classInstance, new Clause[]{
+                    new StringLiteral("Checking if donor has enough blood for a transfusion..."),
```
Comment (Hedgemon4) [inline src/oop/object_interactions/medium/q4/MainTest.java:146]: The "." need to be escaped here as they are special regex characters

File: src/oop/object_interactions/medium/q4/MainTest.java [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+                    new StringLiteral("Blood transfusion completed"),
+                    new Optional(new StringLiteral(" "))
+            });
+            _assertEquals(volume + 1, testClass.getFieldValue(classInstance, volumeAttributeName),
+                    "Your " + transfusionFromMethodName + " method does not correctly add blood to the recipient.");
+            _assertEquals(donVolume - 1, donor.getVolume(),
+                    "Your " + transfusionFromMethodName + " method does not correctly take blood from the donor.");
+
+        }
+        assertNull(output, String.join(" ", "Your", transfusionFromMethodName, "should not return any output"));
```
Comment (Hedgemon4) [inline src/oop/object_interactions/medium/q4/MainTest.java:166]: This is not necessary as you already checked that the return type is void.

File: src/oop/object_interactions/medium/q4/MainTest.java [hunk 22]
```diff
@@ -0,0 +1,220 @@
...
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    @Tag("dependent2")
```
Comment (Hedgemon4) [inline src/oop/object_interactions/medium/q4/MainTest.java:171]: Please fix the tags as I specified in discord.

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/oop/object_interactions/medium/q4/QuestionText.md [hunk 23]
```diff
@@ -0,0 +1,51 @@
+# oop-object_interactions-medium-q4
+
+- Java question
+
+## Question Text
+
+You are working as a vet assistant in an animal hospital and sometimes you encounter an animal that requires a blood 
+transfusion in order to save its life.  Make a class called **Snail** for this project with the following description:
```
Comment (Hedgemon4) [inline src/oop/object_interactions/medium/q4/QuestionText.md:8]: Also, add that they need to code a test class.

File: src/oop/object_interactions/medium/q4/MainTest.java [hunk 24]
```diff
@@ -0,0 +1,216 @@
...
+        Snail donor = new Snail(donType, donVolume);
+        String incorrectMethodDefinition = "The %s method in the %s class is not defined correctly. Make sure it is declared, spelt correctly, and has the correct parameters.";
+        String incorrectModifierMessage = "The %s method in the %s class has the wrong visibility modifier.";
+        String incorrectReturnType = "The %s method in the %s class has the incorrect return type.";
+        assertTrue(testClass.hasMethod(transfusionFromMethodName, new Class<?>[]{Snail.class}), String.format(incorrectMethodDefinition, transfusionFromMethodName, className));
+        assertTrue(testClass.hasModifier(transfusionFromMethodName, new Class<?>[]{Snail.class}, "public"), String.format(incorrectModifierMessage, transfusionFromMethodName, className));
+        assertTrue(testClass.hasReturnType(transfusionFromMethodName, new Class<?>[]{Snail.class}, Void.TYPE), String.format(incorrectReturnType, transfusionFromMethodName, className));
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] methodArguments = {
+                {donor, Snail.class}
```
Comment (Hedgemon4) [inline src/oop/object_interactions/medium/q4/MainTest.java:135]: Also, please do not call the class like this in case they modify something, which will cause out code not to compile. There is a way to get the type using `ObjectTest`

File: src/oop/object_interactions/medium/q4/MainTest.java [hunk 25]
```diff
@@ -0,0 +1,216 @@
...
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    @Tag("dependent1")
+    public void correctTransfusionFromMethod(char type, double volume, char donType, double donVolume) throws Throwable {
+        Object[][] arguments = {
+                {type, char.class},
+                {volume, double.class}
+        };
+        Snail donor = new Snail(donType, donVolume);
```
Comment (Hedgemon4) [inline src/oop/object_interactions/medium/q4/MainTest.java:126]: Please do not initialize a `Snail` object like this. Use `ObjectTest` instead.

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/oop/object_interactions/medium/q4/QuestionText.md [hunk 26]
```diff
@@ -0,0 +1,52 @@
+# oop-object_interactions-medium-q4
+
+- Java question
+
+## Question Text
+
+You are working as a vet assistant in an animal hospital, and sometimes you encounter an animal that requires a blood 
+transfusion in order to save its life.  Make a class called **Snail** for this project, and a class called 
+**TestTransfusion** to test it, with the following description:
```
Comment (Hedgemon4) [inline src/oop/object_interactions/medium/q4/QuestionText.md:9]:
```suggestion
You are working as a vet assistant in an animal hospital, and sometimes you encounter an animal that requires a blood 
transfusion to save its life.  Using the following description, make a class called **Snail** 
to model this situation, and a class called **TestTransfusion** to test your **Snail** class.
```

File: src/oop/object_interactions/medium/q4/QuestionText.md [hunk 27]
```diff
@@ -0,0 +1,52 @@
...
+- Methods:
+    - A `toString` method that prints the object information in the format
+      `My blood type is x and I have yL of blood left` where x is the **bloodType** value and y is the **volume** value.
+    - A `transfusionFrom` method that take another **Snail** object to do a blood tansfer.
+    - It should check if the blood types are a match and if not print `Blood type incompatible`. Note that in this scenario
+  **Snail** blood types are not like humans, so you should check that the type match exactly.
+    - If the types are compatible it should print `Checking if donor has enough blood for a transfusion...`
+    - Then it should make sure that the donor has more than 5 unit of blood if not
+      it should print `Donor does not have enough blood for a transfusion`.
+    - If donor has enough take 1 unit of blood from donor
```
Comment (Hedgemon4) [inline src/oop/object_interactions/medium/q4/QuestionText.md:39]:
```suggestion
    - If the donor has enough take 1 unit of blood from the donor
```

File: src/oop/object_interactions/medium/q4/QuestionText.md [hunk 28]
```diff
@@ -0,0 +1,52 @@
...
+## TestTransfusion Class
+
+Your `TestTransfusion` class must have the following items:
+
+- Main Method:
+    - A `main` method which initializes three **Snail** objects called Sarah, Jones, and Cindy who respectively have
+      blood type A, B, and B, and blood volume of 10.5, 8, and 0.5.
+    - Cindy should try to get blood from first Sarah and then Jones.
+    - Then it should print the information of Sarah, Jones, and Cindy in that order using **toString()** method.
+    - An example of the output would be `"Sarah: " + sarah.toString()`.
```
Comment (Hedgemon4) [inline src/oop/object_interactions/medium/q4/QuestionText.md:51]: You should put an actual sample here, not just the code.

File: src/oop/object_interactions/medium/q4/MainTest.java [hunk 29]
```diff
@@ -0,0 +1,220 @@
...
+        }
+    }
+
+    private static Stream<Arguments> inputProvider() {
+        return Stream.of(
+                Arguments.of('A', 10.3, 'B', 4.9),
+                Arguments.of('A', 15.5, 'A', 7.2),
+                Arguments.of('A', 6.4, 'A', minBlood),
+                Arguments.of('A', 0, 'A', minBlood - 1),
+                Arguments.of('A', 0.5, 'A', minBlood + 1),
```
Comment (Hedgemon4) [inline src/oop/object_interactions/medium/q4/MainTest.java:102]: These are good test cases here 👍

File: src/oop/object_interactions/medium/q4/MainTest.java [hunk 30]
```diff
@@ -0,0 +1,220 @@
...
+                    new Optional(new StringLiteral(" "))
+            });
+        } else if (donVolume <= minBlood) {
+            testClass.callMethod(transfusionFromMethodName, methodArguments, new String[]{"public"}, classInstance, new Clause[]{
+                    new StringLiteral("Checking if donor has enough blood for a transfusion\\.\\.\\."),
+                    new Optional(new StringLiteral(" ")),
+                    new NewLine(),
+                    new StringLiteral("Donor does not have enough blood for a transfusion"),
+                    new Optional(new StringLiteral(" "))
+            });
```
Comment (Hedgemon4) [inline src/oop/object_interactions/medium/q4/MainTest.java:153]: For the ones where no change should occur, you should also test that the blood volume doesn't change.

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/oop/object_interactions/medium/q4/MainTest.java [hunk 31]
```diff
@@ -0,0 +1,224 @@
...
+            _assertEquals(volume, testClass.getFieldValue(classInstance, volumeAttributeName),
+                    "Your " + transfusionFromMethodName + " method does not correctly add blood to the recipient.");
+            _assertEquals(donVolume, testClass.getFieldValue(donor, volumeAttributeName),
+                    "Your " + transfusionFromMethodName + " method does not correctly take blood from the donor.");
```
Comment (Hedgemon4) [inline src/oop/object_interactions/medium/q4/MainTest.java:157]: This should be extended for the type != donType case as well.

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]
